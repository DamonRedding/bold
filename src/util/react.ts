import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
