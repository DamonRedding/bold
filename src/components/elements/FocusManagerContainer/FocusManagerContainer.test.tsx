import { mount, shallow } from 'enzyme'
import * as React from 'react'
import * as waait from 'waait'

import { FocusManagerContainer } from './FocusManagerContainer'

it('should render children', () => {
    expect(shallow(<FocusManagerContainer><span>Children</span></FocusManagerContainer>)).toMatchSnapshot()
})

it('should call onFocusIn prop when focus enter the first element of the container', () => {
    const focusIn = jest.fn()
    const wrapper = mount(
        <FocusManagerContainer onFocusIn={focusIn}>
            <input id='input1' />
            <input id='input2' />
        </FocusManagerContainer>
    )
    expect(focusIn).not.toHaveBeenCalled()
    wrapper.find('#input1').simulate('focus')
    expect(focusIn).toHaveBeenCalledTimes(1)
    wrapper.find('#input2').simulate('focus')
    expect(focusIn).toHaveBeenCalledTimes(1)
})

it('should call onFocusOut prop when focus leaves the container', async () => {
    const focusOut = jest.fn()
    const wrapper = mount(
        <FocusManagerContainer onFocusOut={focusOut}>
            <input id='input1' />
            <input id='input2' />
        </FocusManagerContainer>
    )
    expect(focusOut).not.toHaveBeenCalled()

    wrapper.find('#input1').simulate('focus')
    await waait(0)
    expect(focusOut).not.toHaveBeenCalled()

    wrapper.find('#input2').simulate('focus')
    await waait(0)
    expect(focusOut).not.toHaveBeenCalled()

    wrapper.find('#input2').simulate('blur')
    await waait(0)
    expect(focusOut).toHaveBeenCalledTimes(1)
})
