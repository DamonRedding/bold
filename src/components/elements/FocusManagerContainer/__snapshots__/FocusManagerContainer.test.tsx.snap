// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children 1`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
>
  <span>
    Children
  </span>
</div>
`;
